Snippets for tbDEX Credential Issuance Guide (#1254)
TBD54566975/developer.tbd.website: site/docs/tbdex/issuer/vc-issuance.mdx

## Environment Setup

### Install Express

> We'll use Express to set up the VC issuance service. Follow these instructions to [install Express](https://expressjs.com/en/starter/installing.html).

Reviewer: ```suggestion
This example demonstrates how to use Express to set up the VC issuance service. Follow these instructions to [install Express](https://expressjs.com/en/starter/installing.html).
```

### Install Credentials Package

You'll also need to install `@web5/credentials` to create and issue credentials:

```bash
npm i @web5/credentials
```

### Create a DID

Additionally, to issue credentials, you need a [Decentralized Identifier (DID)](/docs/web5/learn/decentralized-identifiers) which will serve as your Issuer's identity.
If you don't already have one, you'll need to [create a DID](/docs/web5/build/decentralized-identifiers/how-to-create-did).

### Create API Service

> 1. Create a file to serve as your API's entry point (e.g. `main.js`), this is the `entry point` file that was created during the Express setup.

Reviewer: ```suggestion
1. Ensure you have a file, such as `main.js`, to act as your API's entry point, which was created during the Express setup.
```

If the file was already created, we probably shouldn't tell them to Create a file

2. Within the file, import the following:

```js title=main.js
import { api } from './api.js';
```

3. To start your Express API service, add the following:

```js title=main.js
const config = {
  port: 3000, // this can be any port you want to listen on
};

const server = api.listen(config.port, () => {
  log.info(`Server listening on port ${config.port}`);
});
```

> 4. Next, create an `api.js` file, this file will define the core API routes and logic for issuing credentials:

Reviewer: ```suggestion
4. Next, create an `api.js` file. This file will define the core API routes and logic for issuing credentials:
```

```js title=api.js
import express from 'express';
import { VerifiableCredential, Jwt } from '@web5/credentials';
// in the next step we'll cover designing this class
import SanctionsCredential from './sanctionsCredential.js';

const app = express();

// Middleware to parse JSON bodies
app.use(express.json());

// Define additional routes as needed, for example:
app.get('/check-sanctions', async (req, res) => {

>   // we'll implement the API route here

Reviewer: ```suggestion
  // implement the API route here
```

for future reference, in the tbDEX guides, I'd like us to steer away from terms like "we're". These are developers working in professional environments. They aren't completing a tutorial, they are building this for their company.

<Shnip
  snippets={[
    {
      snippetContent: sanctionsCredentialClass,
      language: 'JavaScript',
      title: 'sanctionsCredential.js',
    },
  ]}
/>

In the example above, you provide a class property that stores all the sanctions lists checked against.
You have wide flexibility with what you can store in the credential model class, which means that you're free to include any attributes that may be relevant to the credential you are issuing.

## Implement Issuance

With your web server structure in place and credential designed, you're now ready to build out your issuance codepath.
This is the code you'll execute when a user requests to be issued a VC.

In this code, you'll want to perform any checks required to issue them a credential which will later be verified by an interested third party.

In the example below, we'll assume the user has provided a JWT that contains their identity information (e.g. name, DOB, etc), and your application checks applicable sanctions lists to determine if this user is on them:

<Shnip
  snippets={[
    {
      snippetContent: checkSanctionsEndpoint,
      language: 'JavaScript',
      title: 'api.js',
    },
  ]}
/>

> In order to run this code, lets create a super simple `checkSanctionsLists()` function, you can place this at the top of your `api.js` file:

Reviewer: ```suggestion
In order to run this code, create a `checkSanctionsLists()` function in your `api.js` file:
```

<Shnip
  snippets={[
    {
      snippetContent: checkSacntionsListsFunction,
      language: 'JavaScript',
      title: 'api.js',
    },
  ]}
/>

> Now back in your `"/check-sanctions"` endpoint, we're going to handle the results of the sanction check and create a Verifiable Credential:

Reviewer: ```suggestion
The `/check-sanctions` endpoint should handle the results of the sanction check and create a Verifiable Credential:
```

}
```

To complete the issuance, you must sign the credential.
Signing will return a [VC JSON Web Token](/docs/web5/build/verifiable-credentials/jwt-to-vc#what-is-a-jwt), which is ideal for secure transmission of the credential:

:::note

> The entire PortableDid is used to sign Verifiable Credentials. Please refer to our [Create A DID](/docs/web5/build/decentralized-identifiers/how-to-create-did) Guide for steps on how to create one.

Reviewer: ```suggestion
The BearerDid is used to sign Verifiable Credentials that you issue. Refer to the [Create A DID](/docs/web5/build/decentralized-identifiers/how-to-create-did) guide for steps on how to create one.
```

:::

<Shnip
  snippets={[
    {
      snippetContent: signCreatedSanctionsCredential,
      language: 'JavaScript',
      title: 'api.js',
    },
  ]}
/>

> ## Run Issuance Service

Reviewer: ```suggestion
## Test Issuance Service
```

since they will be running this in production, we should reframe this as just a test
